crates/biome_formatter/src/builders.rs

```diff
@@ -1254,6 +1283,109 @@ pub fn soft_space_or_block_indent<Context>(content: &impl Format<Context>) -> Bl
     }
 }
 
+/// Conditionally adds spaces around the content if its enclosing group fits on a line
+/// _and_ the caller suggests that the space should be added. Otherwise indents the
+/// content and separates it by line breaks.
+///
+/// # Examples
+///
+/// Adds line breaks and indents the content if the enclosing group doesn't fit on the line.
+///
+/// ```
+/// use biome_formatter::{format, format_args, LineWidth, SimpleFormatOptions};
+/// use biome_formatter::prelude::*;
+///
+/// # fn main() -> FormatResult<()> {
+/// let context = SimpleFormatContext::new(SimpleFormatOptions {
+///     line_width: LineWidth::try_from(10).unwrap(),
+///     ..SimpleFormatOptions::default()
+/// });
+///
+/// let elements = format!(context, [
+///     group(&format_args![
+///         text("{"),
+///         soft_block_indent_with_conditional_space(&format_args![
+///             text("aPropertyThatExceeds"),
+///             text(":"),
+///             space(),
+///             text("'line width'"),
+///         ], false),
+///         text("}")
+///     ])
+/// ])?;
+///
+/// assert_eq!(
+///     "{\n\taPropertyThatExceeds: 'line width'\n}",
+///     elements.print()?.as_code()
+/// );
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Adds spaces around the content if the caller requests it and the group fits on the line.
+/// ```
+/// use biome_formatter::{format, format_args};
+/// use biome_formatter::prelude::*;
+///
+/// # fn main() -> FormatResult<()> {
+/// let elements = format!(SimpleFormatContext::default(), [
+///     group(&format_args![
+///         text("{"),
+///         soft_block_indent_with_conditional_space(&format_args![
+///             text("a"),
+///             text(":"),
+///             space(),
+///             text("5"),
+///         ], true),
+///         text("}")
+///     ])
+/// ])?;
+///
+/// assert_eq!(
+///     "{ a: 5 }",
+///     elements.print()?.as_code()
+/// );
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Does not add spaces around the content if the caller denies it and the group fits on the line.
+/// ```
+/// use biome_formatter::{format, format_args};
+/// use biome_formatter::prelude::*;
+///
+/// # fn main() -> FormatResult<()> {
+/// let elements = format!(SimpleFormatContext::default(), [
+///     group(&format_args![
+///         text("{"),
+///         soft_block_indent_with_conditional_space(&format_args![
+///             text("a"),
+///             text(":"),
+///             space(),
+///             text("5"),
+///         ], false),
+///         text("}")
+///     ])
+/// ])?;
+///
+/// assert_eq!(
+///     "{a: 5}",
+///     elements.print()?.as_code()
+/// );
+/// # Ok(())
+/// # }
+/// ```
+pub fn soft_block_indent_with_conditional_space<Context>(
+    content: &impl Format<Context>,
+    should_add_space: bool,
+) -> BlockIndent<Context> {
+    if should_add_space {
+        soft_space_or_block_indent(content)
+    } else {
+        soft_block_indent(content)
+    }
+}
+
```

Comment:
This is similar to the `conditional_space`, but a little more involved? I didn't have a super obvious gut feeling on how to go with this, so I decided to make it "generic" where the condition for whether or not the space is added gets passed in by the caller.